speculos/api/api.py

```diff
@@ -1,36 +1,42 @@
 import socket
 import threading
 import pkg_resources
-from typing import Dict, Optional
+from typing import Any, Dict
 from flask import Flask
 from flask_restful import Api
 
-from ..mcu.readerror import ReadError
-from ..mcu.seproxyhal import SeProxyHal
+from speculos.mcu.display import DisplayNotifier, IODevice
+from speculos.mcu.readerror import ReadError
+from speculos.mcu.seproxyhal import SeProxyHal
+from speculos.observer import BroadcastInterface
 from .apdu import APDU
 from .automation import Automation
 from .button import Button
-from .events import Events, EventsBroadcaster
+from .events import Events
 from .finger import Finger
 from .screenshot import Screenshot
 from .swagger import Swagger
 from .web_interface import WebInterface
 from .ticker import Ticker
 
 
-class ApiRunner:
-    """Run the Speculos API server, with a notification when it stops"""
+class ApiRunner(IODevice):
+    """Run the Speculos API server in a dedicated thread, with a notification when it stops"""
     def __init__(self, api_port: int) -> None:
-        self._app: Optional[Flask] = None
-        # self.s is used by Screen.add_notifier. Closing self._notify_exit
+        self._app: Flask
+        # self.sock is used by Screen.add_notifier. Closing self._notify_exit
         # signals it that the API is no longer running.
-        self.s: socket.socket
+        self.sock: socket.socket
         self._notify_exit: socket.socket
-        self.s, self._notify_exit = socket.socketpair()
+        self.sock, self._notify_exit = socket.socketpair()
         self.api_port: int = api_port
 
-    def can_read(self, s: int, screen) -> None:
-        assert s == self.s.fileno()
+    @property
+    def file(self):
+        return self.sock
```

Comment:
I feel like `sock` is more explicit ?

Reply:
In this case yes, but it is an concrete implementation of the abstract `IODevice.file` method.
`IODevice` itself could have been renamed as a more precise `SocketDevice`, as **all but one** children use sockets.
Only the `VNC` class does not use a socket, but a `subprocess.PIPE`. That's why I chose to widen the `socket` to a more generic `file`. :shrug: